new module to create cloudwatch log group

willricardo opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Added a module cloudwatchlogs_log_group to create or delete an Aws CloudWatch Log group
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
cloudwatchlogs_log_group
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (test 85116f88ec) last updated 2017/10/27 16:16:39 (GMT -200)
  config file = None
  configured module search path = ['/home/willricardo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /arq/app/ansible/guru-ansible/ansible/lib/ansible
  executable location = /arq/app/ansible/guru-ansible/ansible/bin/ansible
  python version = 3.5.3 (default, Apr 23 2017, 22:31:59) [GCC 4.8.5]
```

willricardo committed:
new module to create cloudwatch log group

willricardo committed:
input RETURN

willthames commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,149 @@
+
+
+def delete_log_group(client, log_group_name):
+    desc_log_group = client.describe_log_groups(logGroupNamePrefix=log_group_name)
```
all calls to AWS should have [exception handling](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#exception-handling-for-boto3-and-botocore)

willricardo replied:
Opss.. Sorry!!
I fix this error.

willthames commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,149 @@
+    if tags:
+        request['tags'] = tags
+
+    client.create_log_group(**request)
```
Needs exception handling

willthames replied:
I know this returns nothing, but it's still worth returning some information to the caller, perhaps using
http://boto3.readthedocs.io/en/latest/reference/services/logs.html#CloudWatchLogs.Client.describe_log_groups

willricardo replied:
Really, do you have reason, this is better.
I put the return with base in describe_logs.

willricardo replied:
I put the exception handling and include the retention policy too.

willthames reviewed (commented):
Is it worth handling retention policy too?
http://boto3.readthedocs.io/en/latest/reference/services/logs.html#CloudWatchLogs.Client.put_destination_policy

willthames commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,149 @@
+    state = module.params.get('state')
+
+    if state == 'present':
+        create_log_group(client=logs, log_group_name=module.params['log_group_name'], kms_key_id=module.params['kms_key_id'], tags=module.params['tags'])
```
What happens if the log group already exists?

What happens if someone wants to change the tags of the logs group?

willricardo replied:
The ResourceAlreadyExistsException was raised, if the log group already existed.
But I changed this to return the describe_logs information if the log group already exist, and the execution do not failure and ensure idempotency.

I put more two funcionalities.:
a)  overwrite, if the user wants override the log group.
b)  state = get . If the user wants list log group.

I guess to handle log group tags , should be another module.
Maybe a cloudwatchlogs_tag with these funcionalities.:
a) tag_log_group()  = state present
b) untag_log_group() = state absent
c) list_tags_log_group() = state get

If you agree, I will write this module

willthames reviewed (dismissed):
Exception handling and idempotency require addressing.

willricardo committed:
put exception handling and ensure idempotency

willricardo commented:
ready_for_review

mixja commented:
I would recommend validating the retention time in days, or at the very least referencing the valid values in the docs - see http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutRetentionPolicy.html

mixja commented:
Also I would consider adding support for log subscriptions - see http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutSubscriptionFilter.html

willricardo committed:
include check to valid values in put_retention_policy

willricardo committed:
include valid values to retention policy in documentation

Remove state=get from cloudwatchlogs_log_group and create a cloudwatchlogs_log_group_facts module

Handle BotoCoreErrors

Use camel_dict_to_snake_dict instead of _camel_to_snake

Fix idempotence and simplify logic by checking if the log group exists first

willricardo commented:
included validation to valid values in retention policy.
About the support for log subscriptions, i think that is better create another module to do this.
I'd like create this module too to support all features to log subscription.

willricardo commented:
ready_for_review

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+        delete_log_group(client=logs,
+                         log_group_name=module.params['log_group_name'],
+                         module=module)
+    elif state == 'get':
```
This logic should really go in a cloudwatchlogs_log_group_facts.py module. You can move any of the shared code - maybe describe_log_group() - into module_utils/aws/.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+        desc_log_group_result = dict(changed=False, stdout=final_log_group_snake)
+        module.exit_json(**desc_log_group_result)
+
+    else:
```
You can remove this else because the ec2_argument_spec enforces that one of the allowed choices is provided.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+
+import traceback
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import _camel_to_snake, camel_dict_to_snake_dict, boto3_conn, ec2_argument_spec, get_aws_connection_info
```
I'm not sure _camel_to_snake is stable to import and use. Are there cases when camel_dict_to_snake_dict isn't working right?

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+
+
+def create_log_group(client, log_group_name, kms_key_id, tags, retention, module):
+    try:
```
You could move this try/except just to be around the create_log_group call.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+                else:
+                    module.fail_json(msg="The aws CloudWatchLogs log group was not created. \n please try again!")
+
+    except botocore.exceptions.ClientError as e:
```
This should probably be: `except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:` as noted in the [guidelines ](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-1). Same for the other try/excepts.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+                    module.fail_json(msg="The aws CloudWatchLogs log group was not created. \n please try again!")
+
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg=e.response, exception=traceback.format_exc(),
```
It's preferable to have msg to be something specific short and specific about what call failed - such as "Unable to create log group: {0}".format(log_group_name) since e.response is being used twice. Same for any following too.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+        overwrite=dict(required=False, type='bool', default=False)
+    ))
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
```
This module does not support check mode. If running a playbook with --check, it should show changed: true or false in the output but not actually make said changes.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+        if not region:
+            module.fail_json(msg="Region must be specified as a parameter, in EC2_REGION or AWS_REGION environment variables or in boto configuration file")
+        logs = boto3_conn(module, conn_type='client', resource='logs', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.ProfileNotFound as e:
```
You don't need to wrap the connection in try/except, it is now handled in module_utils/ec2 (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-1).

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudwatchlogs_log_group.py:
```diff
@@ -0,0 +1,298 @@
+
+
+def delete_log_group(client, log_group_name, module):
+    try:
```
You can just have the try/except around the delete_log_group() section since describe_log_group() has its own handling.

s-hertel reviewed (commented):
This looks really good - some minor comments, and wondering if logic for state 'get' should be moved into a facts module instead.

s-hertel commented:
I have opened https://github.com/ansible/ansible/pull/35314 for the facts module.

s-hertel committed:
Add new module cloudfrontwatchlogs_log_group to the changelog.
